### PR TITLE
Spark file dataobject incremental mode (#517)

### LINKED ISSUE #517: Incremental processing with SparkFileDataObjects
**Is your feature request related to a problem? Please describe.**
I would like to process input data of SparkFileDataObjects incrementally and efficient.

**Describe the solution you'd like**
Observe filenames processed by SparkFileDataObjects.getSparkDataFrame and delete them with FileIncrementalMoveMode.
In addition to delete the files processed, it would be nice to optionally configure to archive them in a subfolder. 
And we could also implement CanCreateIncrementalOutput for SparkFileDataObjects to enable DataObjectStateIncrementalMode for SparkFileDataObjects, see https://spark.apache.org/docs/latest/sql-data-sources-generic-options.html#modification-time-path-filters

**Additional context**
FileIncrementalMoveMode is currently only supported with FileSubFeeds.

### PR DESCRIPTION
**Pull request by zzeekk**
### What changes are included in the pull request?
- implement FileIncrementalMoveMode for SparkFileDataObject by observing files processed
- implement FileIncrementalMoveMode archiveSubdirectory parameter
- implement CanCreateIncrementalOutput for SparkFileDataObject

### Why are the changes needed?
see #517

### CODE REVIEW

File: sdl-core/src/test/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObjectTest.scala [hunk 1]
```diff
@@ -402,4 +402,39 @@ class SparkFileDataObjectTest extends DataObjectTestSuite with SmartDataLakeLogg
     assert(compactedTstmp == compactedTstmp2)
   }
 
+
+  test("incremental output mode") {
+
+    // create data object
+    val tempDir = Files.createTempDirectory("tempHadoopDO")
+    val dataObject = ParquetFileDataObject(id = "partitionTest", path = tempDir.toString, saveMode = SDLSaveMode.Append)
+    //dataObject.filesystem.setWriteChecksum(false)
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObjectTest.scala]: remove commented out coe

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/Observation.scala [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+package io.smartdatalake.util.spark
+
+import org.apache.spark.sql._
+import org.apache.spark.sql.execution.QueryExecution
+import org.apache.spark.sql.util.QueryExecutionListener
+
+import java.util.UUID
+
+/**
+ * This code is copied from Spark 3.3.0 when it was not yet released (and simplified).
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/spark/Observation.scala]: Well... in that case ut should be moved to the other repo spark-extensions, right?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala [hunk 3]
```diff
@@ -147,15 +149,57 @@ case class JdbcTableDataObject(override val id: DataObjectId,
...
     val queryOrTable = Map(table.query.map(q => ("query",q)).getOrElse("dbtable"->table.fullName))
-    val df = context.sparkSession.read.format("jdbc")
+    var df = context.sparkSession.read.format("jdbc")
       .options(options)
       .options(connection.getAuthModeSparkOptions)
       .options(queryOrTable)
       .load()
+    incrementalOutputState.foreach { case (lastColumn, lastWatermark)  =>
+      assert(incrementalOutputColumn.isDefined)
+      if (lastColumn != incrementalOutputColumn.get) logger.warn(s"($id) incrementalOutputState has different column as incrementalOutputColumn ($lastColumn != ${incrementalOutputColumn.get}")
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala:160]: message confuses me. incrementalOutputState is lastColumn ?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala [hunk 4]
```diff
@@ -477,6 +521,10 @@ private[smartdatalake] case class JdbcColumn(name: String, isNameCaseSensitiv: B
   def nameEqualsIgnoreCaseSensitive(name: String): Boolean = {
     this.name.equalsIgnoreCase(name)
   }
+  def getSparkField(schema: StructType): Option[StructField] = {
+    if (isNameCaseSensitiv) schema.find(_.name.toLowerCase == name.toLowerCase)
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: isNameCaseSensitive

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObject.scala [hunk 5]
```diff
@@ -122,17 +125,31 @@ private[smartdatalake] trait SparkFileDataObject extends HadoopFileDataObject
       filesystem.mkdirs(hadoopPath)
     }
 
+    val incrementalOutputOptions = if (incrementalOutputState.isDefined && context.phase == ExecutionPhase.Exec) {
+      val previousOutputState = incrementalOutputState.get
+      // Comparison of modifiedAfter and modifiedBefore are both exclusive on Microsecond level, but file timestamps maximum detail is milliseconds.
+      // Actually comparison of one operator should be inclusive to avoid reading files in edge cases.
+      // Current timestamp is also at millisecond level. If we subtract one microsecond from current timestamp we can avoid the problems because of exclusive comparison.
+      incrementalOutputState = Some(LocalDateTime.now.minusNanos(1000))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObject.scala:133]: Do you have a test for that? It's a bit hard to follow, an example would help. So you are making sure modifiedBefore is always smaller than modified after?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala [hunk 6]
```diff
@@ -208,7 +208,10 @@ private[smartdatalake] case class ActionDAGRun(dag: DAG[Action], executionId: SD
    */
   def getRuntimeInfos: Map[ActionId, RuntimeInfo] = {
     dag.getNodes.map { a =>
-      (a.id, a.getRuntimeInfo(Some(executionId)).getOrElse(RuntimeInfo(executionId, RuntimeEventState.PENDING)))
+      var runtimeInfo = a.getRuntimeInfo(Some(executionId)).getOrElse(RuntimeInfo(executionId, RuntimeEventState.PENDING))
+      // overwrite DataObjectsState for skipped actions to keep previous state.
+      if (runtimeInfo.state == RuntimeEventState.SKIPPED) runtimeInfo = runtimeInfo.copy(dataObjectsState = initialDataObjectsState)
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:213]: why do you need to copy it?

File: sdl-core/src/test/scala/io/smartdatalake/app/SmartDataLakeBuilderTest.scala [hunk 7]
```diff
@@ -491,13 +491,22 @@ class SmartDataLakeBuilderTest extends FunSuite with BeforeAndAfter {
...
     action1.reset
     sdlb.run(sdlConfig)
 
     // check results
     val dfResult3 = tgt1DO.getSparkDataFrame(Seq())
-    assert(dfResult3.select(functions.max($"nb".cast("int")), functions.count("*")).as[(Int,Long)].head == (30,30))
+    assert(dfResult3.select(functions.max($"nb".cast("int")), functions.count("*")).as[(Int,Long)].head == (20,20))
+
+    // start 4th dag run
+    // TODO: keep state when skipped!
```
Comment (Geheiner) [inline sdl-core/src/test/scala/io/smartdatalake/app/SmartDataLakeBuilderTest.scala]: you have already implemented this TODO in io.smartdatalake.workflow.ActionDAGRun#getRuntimeInfos

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/Observation.scala (hunk 2 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/spark/Observation.scala]: It has a similar design, but is no longer exactly the same. If fixed the comment from "copied" to "inspired", and would leave it in SDLB.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala (hunk 3 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala:160]: The incrementalOutputColumn configuration options is used to initialize State in the first run. SDLB keeps the column also as part of the state. In case the configuration option should change in the future, there is at least a warn message...

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: I will remove method getSparkField completely... it is a left-over from an earlier approach.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala (hunk 6 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:213]: Well, i want to update a field of the case class. How would you do this?

File: sdl-core/src/test/scala/io/smartdatalake/app/SmartDataLakeBuilderTest.scala (hunk 7 above)
Comment (zzeekk) [reply sdl-core/src/test/scala/io/smartdatalake/app/SmartDataLakeBuilderTest.scala]: Good catch!

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObject.scala (hunk 5 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/SparkFileDataObject.scala:133]: I making sure that modifiedBefore/After can not be the same as a files modified date, because file modified date has only milliseconds resolution.
Like that i can work around the fact, that modifiedBefore/After are both exclusive conditions, e.g. "greater than" and not "greater or equal than", but i would need one of them to be inclusive to not lose some data.
I test this in SparkFileDataObject "incremental output mode".

File: sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala (hunk 6 above)
Comment (Geheiner) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/ActionDAGRun.scala:213]: ah of course, I forgot about that scala pattern

**Review by Geheiner — APPROVED** [review]
LGTM,I'm curious to see the real world example on which you will be using this

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala [hunk 8]
```diff
@@ -161,7 +161,8 @@ case class JdbcTableDataObject(override val id: DataObjectId,
       val newDataType = if (SchemaUtil.isSparkCaseSensitive) df.schema.find(_.name == incrementalOutputColumn.get).get.dataType
       else df.schema.find(_.name.equalsIgnoreCase(incrementalOutputColumn.get)).get.dataType
       if (context.phase == ExecutionPhase.Exec) {
-        val newWatermarkValue = df.agg(max(col(incrementalOutputColumn.get))).head.get(0)
+        val newWatermarkValue = Option(df.agg(max(col(incrementalOutputColumn.get))).head.get(0))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: @zzeekk 
watermark is probably not the right word for this.
Definition form google: Software watermarking involves embedding a unique identifier within a piece of software, to discourage software theft. 
Maybe bookmark would be better? Or readPointer? Perhaps we could brainstorm this

Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: @Geheiner right, but it's used else where in streaming, e.g. Kafka: https://docs.confluent.io/platform/current/clients/confluent-kafka-dotnet/_site/api/Confluent.Kafka.WatermarkOffsets.html
I think it's more clear if i call it "highWatermark" (which translates to Hochwasserstand) and dont simplify it to watermark. WDJT?

Comment (pgruetter) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: @zzeekk @Geheiner I would have proposed highWatermark as well. The term is also used internally by RDMSs.

Comment (Geheiner) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala]: ahh they  are talking watermarks in the context of "Wasserpegelstand" and not "Wasserzeichen".
"A high water mark is a point that represents the maximum rise of a body of water over land. Such a mark is often the result of a flood, but high water marks may reflect an all-time high, an annual high or the high point for some other division of time."
Then I agree with the two of you, highWatermark sounds good!
